Update developer-guide for `@now/build-utils` (#661)
zeit/docs: pages/docs/v2/deployments/builders/developer-guide.mdx

  description:
    'How to create a Builder to take ZEIT Now deployment sources and output either static files or lambdas with almost limitless possibilities.',
  editUrl: 'pages/docs/v2/deployments/builders/developer-guide.mdx',
  lastEdited: '2019-03-22T22:53:49.000Z'
}

A builder is an npm module that exposes a `build` function and optionally an `analyze` function and `prepareCache` function.

In short:

- The `analyze` function returns a unique fingerprint used for the purpose of [build de-duplication](/docs/v2/deployments/concepts/immutability). If the `analyze` function is not supplied, it's equivalent to it returning a random fingerprint for each build.

> - The `build` function returns a [Files](#files) datastructure that contains the [build outputs](/docs/v2/deployments/builds#sources-and-outputs), which can be:

Reviewer: 'datastructure'
=>
'data structure'

  - Static files
  - [Lambda](/docs/v2/deployments/concepts/lambdas) functions
- The `prepareCache` function is equivalent to `build`, but it executes the instructions necessary to prepare a cache for the next run.

The arguments these functions receive are:

> - `files`: All source files of the project as a [Files](#files) datastructure

Reviewer: 'All source files of the project as a [Files](#files) datastructure'
=>
'All the source files of a project as a [Files](#files) data structure.'

  description:
    'How to create a Builder to take ZEIT Now deployment sources and output either static files or lambdas with almost limitless possibilities.',
  editUrl: 'pages/docs/v2/deployments/builders/developer-guide.mdx',
  lastEdited: '2019-03-25T12:58:45.000Z'
}

A builder is an npm module that exposes a `build` function and optionally an `analyze` function and `prepareCache` function.

In short:

> - The `analyze` function returns a unique fingerprint used for the purpose of [build de-duplication](/docs/v2/deployments/concepts/immutability). If the `analyze` function is not supplied, it's equivalent to it returning a random fingerprint for each build.

Reviewer: L15 => L19
Can we look at breaking these up into 3 sections with headers? I find the amount of text makes it a little more difficult to read than it needs to be.
We could use the headers as anchors for other content to easily redirect here as well.
Also, putting the function signatures in their respective section would make it easier to digest.

Author: Done!

- The `build` function returns a [Files](#files) data structure that contains the [build outputs](/docs/v2/deployments/builds#sources-and-outputs), which can be:
  - Static files
  - [Lambda](/docs/v2/deployments/concepts/lambdas) functions
- The `prepareCache` function is equivalent to `build`, but it executes the instructions necessary to prepare a cache for the next run.

The arguments these functions receive are:

- `files`: All source files of the project as a [Files](#files) data structure
- `entrypoint`: Name of entrypoint file for this particular build job. Value `files[entrypoint]` is guaranteed to exist and be a valid [File](#files) reference. `entrypoint` is always a discrete file and never a glob, since globs are expanded into separate builds at deployment time.
- `workPath`, a writable temporary directory where you are encouraged to perform your build process. This directory will be populated with the restored cache from the previous run (if any) for `analyze` and `build`.
- `cachePath`, a writable temporary directory where you can build a cache for the next run. This is only passed to `prepareCache`

> - `config`, an arbitrary object passed from by the user in the [build definition](#defining-the-build-step) in `now.json`.

Reviewer: L23 => L27
There is a mix of both commas and colons used after the `argument` - can we change this to be just colons instead?

Author: Nice catch!

  description:
    'How to create a Builder to take ZEIT Now deployment sources and output either static files or lambdas with almost limitless possibilities.',
  editUrl: 'pages/docs/v2/deployments/builders/developer-guide.mdx',
  lastEdited: '2019-03-25T15:03:41.000Z'
}

A builder is an npm module that exposes a `build` function and optionally an `analyze` function and `prepareCache` function.

> Typically, builders are published to [npmjs.com](https://npmjs.com) as a package and referenced in the `use` property of the `now.json` configuration file.

Reviewer 2: I don't think we can use "Typically" here. There's no other possibility.

Reviewer 2: Also Builders is capitalised.

Author: I haven't confirmed, but I heard that builders do `npm install` on the string so it should work with github urls, etc. I should confirm this first.

Reviewer 3: Well technically you could use a URL, or any other valid `npm` install specifier syntax (see https://docs.npmjs.com/files/package.json#repository)

Reviewer 2: Then this isn't specific enough. 

"Builders are modules referenced through the `package.json` dependency value syntax"

Or something alike

Author: Ok, so I added another sentence to clean this up and linked to the npm install docs since thats what would be used (the repository docs that Nate linked to are not the same thing).

> See the [Builders Overview](/v2/deployments/builders/overview) to see builder usage.

Reviewer 2: ```suggestion
See the [Builders Overview](/v2/deployments/builders/overview) to see Builder usage.
```

## Builder Exports

> ### analyze

Reviewer 2: ```suggestion
### `analyze`
```

> An **optional** exported function that returns a unique fingerprint used for the purpose of [build de-duplication](/docs/v2/deployments/concepts/immutability). If the `analyze` function is not supplied, it's equivalent to it returning a random fingerprint for each build.

Reviewer 2: ```suggestion
An **optional** exported function that returns a unique fingerprint used for the purpose of [build de-duplication](/docs/v2/deployments/concepts/immutability#deduplication-algorithm). If the `analyze` function is not supplied, a random fingerprint is assigned to each build.
```

```js
export analyze({
  files: Files,
  entrypoint: String,
  workPath: String,
  config: Object

> }) : String fingerprint

Reviewer 3: This wasn't introduced by your PR. But it bothers me that there's inconsistent spacing on the `:` for the return type (as compared to the `prepareCache()` function for example).

```

If you are using TypeScript, you should use the following types:

```ts
import { AnalyzeOptions } from '@now/build-utils'
export analyze(options: AnalyzeOptions) {
  return 'fingerprint goes here'
}
```

> ### build

Reviewer 2: ```suggestion
### `build`
```

> A **required** exported function that returns a [Files](#files) data structure that contains the [build outputs](/docs/v2/deployments/builds#sources-and-outputs), which can be a [Static File](#file) or a [Lambda](#lambda).

Reviewer 2: ```suggestion
A **required** exported function that returns a [Files](#files) data structure that contains the [Build outputs](/docs/v2/deployments/builds#sources-and-outputs), which can be a [Static File](#file) or a [Lambda](#lambda).
```

  description:
    'How to create a Builder to take ZEIT Now deployment sources and output either static files or lambdas with almost limitless possibilities.',
  editUrl: 'pages/docs/v2/deployments/builders/developer-guide.mdx',
  lastEdited: '2019-03-25T19:20:59.000Z'
}

A builder is an npm module that exposes a `build` function and optionally an `analyze` function and `prepareCache` function.
Official Builders are published to [npmjs.com](https://npmjs.com) as a package and referenced in the `use` property of the `now.json` configuration file.

> However, the `use` property will work with any [npm install argument](https://docs.npmjs.com/cli/install) such as a git repo url which is useful for testing your builder.

Reviewer 2: ```suggestion
However, the `use` property will work with any [npm install argument](https://docs.npmjs.com/cli/install) such as a git repo url which is useful for testing your Builder.
```

  description:
    'How to create a Builder to take ZEIT Now deployment sources and output either static files or lambdas with almost limitless possibilities.',
  editUrl: 'pages/docs/v2/deployments/builders/developer-guide.mdx',
  lastEdited: '2019-03-25T12:58:45.000Z'
}

> A builder is an npm module that exposes a `build` function and optionally an `analyze` function and `prepareCache` function.

Reviewer: 'A builder is an npm module that exposes a `build` function and optionally an `analyze` function and `prepareCache` function.'
=>
'A builder is an npm module that exposes a `build` function and optionally both an `analyze` and `prepareCache` function.'

Author: This makes it sounds like both `analyze` and `prepareCache` must be used together or none at all when in fact they are both optional.

Author: I added the `optional` or `required` text to each new section. I think this will help the reader understand.

Reviewer: Agreed, I like that approach better.

}) : Files output
```

If you are using TypeScript, you should use the following types:

```ts
import { BuildOptions } from '@now/build-utils'
export build(options: BuildOptions) {
  return { 'path-to-file': File }
}
```

> ### prepareCache

Reviewer 2: ```suggestion
### `prepareCache`
```

> An **optional** exported function that is equivalent to `build`, but it executes the instructions necessary to prepare a cache for the next run.

Reviewer 2: ```suggestion
An **optional** exported function that is equivalent to [`build`](#build), but it executes the instructions necessary to prepare a cache for the next run.
```

}) : Files output
```

If you are using TypeScript, you should use the following types:

```ts
import { BuildOptions } from '@now/build-utils'
export build(options: BuildOptions) {
  return { 'path-to-file': File }
}
```

### `prepareCache`

An **optional** exported function that is equivalent to [`build`](#build), but it executes the instructions necessary to prepare a cache for the next run.

```js
prepareCache({
  files: Files,
  entrypoint: String,
  workPath: String,
  cachePath: String,
  config: Object

> }) : Files cacheOutput

Reviewer 2: Does this line make sense?

Author: I mean it's not valid JS or TS but its what was there before so I didn't really touch it. Its more like a pseudo-code.

Reviewer 3: I'd prefer no space before the `:` personally.

Reviewer 2: Do we prefer to have it there at all, instead of a Caption or something that explains it? I wouldn't understand this if I looked at it for the first time.

}): Files cacheOutput
```

If you are using TypeScript, you can import the types for each of these functions by using the following:

```ts
import { PrepareCacheOptions } from '@now/build-utils'
export prepareCache(options: PrepareCacheOptions) {
  return { 'path-to-file': File }
}
```

### Builder Options

> The exported functions `analyze`, `build`, and `prepareCache` receive one argument with the following properties:

Reviewer 2: ```suggestion
The exported functions [`analyze`](#analyse), [`build`](#build), and [`prepareCache`](#preparecache) receive one argument with the following properties:
```

> - `files`: All source files of the project as a [Files](#files) data structure

Reviewer 3: Inconsistent `.` usage on these bullet points (some end with `.`, some don't).

Reviewer 2: Let's end with `.`

> - `entrypoint`: Name of entrypoint file for this particular build job. Value `files[entrypoint]` is guaranteed to exist and be a valid [File](#files) reference. `entrypoint` is always a discrete file and never a glob, since globs are expanded into separate builds at deployment time.

Reviewer 2: Any link for `entrypoint`?

Reviewer 2: ping on this

> - `workPath`: A writable temporary directory where you are encouraged to perform your build process. This directory will be populated with the restored cache from the previous run (if any) for `analyze` and `build`.

Reviewer 2: ```suggestion
- `workPath`: A writable temporary directory where you are encouraged to perform your build process. This directory will be populated with the restored cache from the previous run (if any) for [`analyze`](#analyse) and [`build`](#build).
```

- `cachePath`: A writable temporary directory where you can build a cache for the next run. This is only passed to `prepareCache`

> - `config`: An arbitrary object passed from by the user in the [build definition](#defining-the-build-step) in `now.json`.

Reviewer 2: ```suggestion
- `config`: An arbitrary object passed from by the user in the [Build definition](#defining-the-build-step) in `now.json`.
```

A [lambda](/docs/v2/deployments/concepts/lambdas) is created where the builder logic is executed. The lambda is run using the Node.js 8 runtime. A brand new sandbox is created for each deployment, for security reasons. The sandbox is cleaned up between executions to ensure no lingering temporary files are shared from build to build.

> All the APIs you export (`analyze`, `build` and `prepareCache`) are not guaranteed to be run in the same process, but the filesystem we expose (e.g.: `workPath` and the results of calling [`getWriteableDirectory`](#getWriteableDirectory) ) is retained.

Reviewer 2: ```suggestion
All the APIs you export ([`analyze`](#analyze), [`build`](#build) and [`prepareCache`](#preparecache)) are not guaranteed to be run in the same process, but the filesystem we expose (e.g.: `workPath` and the results of calling [`getWriteableDirectory`](#getWriteableDirectory) ) is retained.
```

Please test that these links are correct, by the way. I'm just going off of the name

## Types

> 

Reviewer 2: Since we're removing the information of the properties, is this given elsewhere?

Author: Kind of. It's defined in TS below `type Files = { [filePath: string]: File }`

**Methods:**

- `static async fromStream({ mode : Number, stream : Stream }) :FileBlob` creates an instance of a [FileBlob](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object) from [`Stream`](https://nodejs.org/api/stream.html) with `mode`
- `toStream() : Stream` creates a [Stream](https://nodejs.org/api/stream.html) of the file body

### Lambda

```ts
import { Lambda } from '@now/build-utils'
// Deprecated usage: `@now/build-utils/lambda`
```

> This is a [JavaScript class](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes) that can be created by supplying `files`, `handler` and `runtime` and `environment` as an object to the [`createLambda`](#createlambda) helper. The instances of this class should not be created directly. Instead use a call to [`createLambda`](#createlambda).

Reviewer: 'created by supplying `files`, `handler` and `runtime` and `environment` as an object'
=>
'created by supplying `files`, `handler`, `runtime` and `environment` as an object'

The following is exposed by `@now/build-utils` to simplify the process of writing Builders, manipulating the file system, using the above types, etc.

> 

Reviewer 2: Please make sure that any links to these headings are also changed

Author: I checked and there are no other pages linking here

> ### createLambda

Reviewer 2: ```suggestion
### `createLambda`
```

})
```

### download

Signature: `download() : Files`

```ts
import { download } from '@now/build-utils'
// Deprecated usage: `@now/build-utils/fs/download`
```

> This utility allows you to download the contents of a `Files` datastructure, therefore creating the filesystem represented in it.

Reviewer: 'datastructure'
=>
'data structure'

})
```

> ### download

Reviewer 2: ```suggestion
### `download`
```

Signature: `download() : Files`

```ts
import { download } from '@now/build-utils'
```

This utility allows you to download the contents of a `Files` data structure, therefore creating the filesystem represented in it.

Since `Files` is an abstract way of representing files, you can think of `download` as a way of making that virtual filesystem _real_.

```js
await download(files, workPath)
```

> ### glob

Reviewer 2: ```suggestion
### `glob`
```

Signature: `glob() : Files`

```ts
import { glob } from '@now/build-utils'
```

> This utility allows you to _scan_ the filesystem and return a `Files` representation of the matched glob search string. It can be thought of as the reverse of [`download`](#download).

Reviewer 2: ```suggestion
This utility allows you to _scan_ the filesystem and return a [`Files`](#files) representation of the matched glob search string. It can be thought of as the reverse of [`download`](#download).
```

> The following trivial example downloads everything to the filesystem, only to return it back (therefore just re-creating the passed-in `files`):

Reviewer 2: ```suggestion
The following trivial example downloads everything to the filesystem, only to return it back (therefore just re-creating the passed-in [`Files`](#files)):
```

```js
const { glob, download } = require('@now/build-utils')

exports.build = ({ files, workPath }) => {
  await download(files, workPath)
  return glob('**', workPath)
}
```

> ### getWriteableDirectory

Reviewer 2: ```suggestion
### `getWriteableDirectory`
```

Signature: `getWriteableDirectory() : String`

```ts
import { getWriteableDirectory } from '@now/build-utils'
```

In some occasions, you might want to write to a temporary directory.

> ### rename

Reviewer 2: ```suggestion
### `rename`
```

})
```

### `download`

Signature: `download() : Files`

```ts
import { download } from '@now/build-utils'
```

> This utility allows you to download the contents of a `Files` data structure, therefore creating the filesystem represented in it.

Reviewer 2: ```suggestion
This utility allows you to download the contents of a [`Files`](#files) data structure, therefore creating the filesystem represented in it.
```
